censustree/censustree.go

```diff
@@ -2,28 +2,33 @@ package censustree
 
 import (
 	"fmt"
-	"path/filepath"
-	"strings"
+	"sync"
 	"sync/atomic"
 
 	"github.com/vocdoni/arbo"
 	"go.vocdoni.io/dvote/db"
-	"go.vocdoni.io/dvote/db/badgerdb"
+	"go.vocdoni.io/dvote/db/prefixeddb"
 	"go.vocdoni.io/dvote/tree"
 	"go.vocdoni.io/proto/build/go/models"
 )
 
 // Tree implements the Merkle Tree used for census
+// Concurrent updates to the tree.Tree can lead to losing some of the updates,
+// so we don't expose the tree.Tree directly, and lock it in every method that
+// updates it.
 type Tree struct {
-	*tree.Tree
+	tree *tree.Tree
 
+	sync.Mutex
```

Comment:
arbo tree already has a mutex for the write operations, maybe this Tree wrapper over arbo does not need it? (and consequently the new methods of the wrapper are not needed?)

Reply:
When I made this PR I didn't consider that Arbo already had mutex for write operations.  Nevertheless I've reviewed the way we are using it and we are creating the WriteTx outside the mutex, which could lead to conflicts, so I think the mutex I added is necessary.  It's not ideal because then we have double locking, but as we discussed AFK, this could be addressed in Arbo by having the FooWithTx functions not lock, and Foo functions lock.  I've added a badgerdb test to validate what happens on concurrent WriteTx that write to the same key.